module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,316 @@
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.validation.BeanPropertyBindingResult;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import javax.validation.Valid;
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Optional;
+
+import static com.epam.esm.constants.AppConstants.DEFAULT_PAGE_NUMBER;
+import static com.epam.esm.constants.AppConstants.DEFAULT_PAGE_SIZE;
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
+
+/**
+ * @author Sergei Kristev
+ * <p>
+ * Gets data from rest in JSON format on path "/gift-certificates".
+ */
+@RestController
+@RequestMapping("/gift-certificates")
+public class GiftCertificateController {
+
+    private final GiftCertificateService giftCertificateService;
+    private final CertificateSearchValidator searchValidator;
+
+    /**
+     * Constructor accepts service layer objects and certificate validator.
+     *
+     * @param giftCertificateService GiftCertificateService instance.
+     * @param searchValidator        CertificateSearchValidator instance.
+     */
+    @Autowired
+    public GiftCertificateController(GiftCertificateService giftCertificateService,
+                                     CertificateSearchValidator searchValidator) {
+        this.giftCertificateService = giftCertificateService;
+        this.searchValidator = searchValidator;
+    }
+
+    /**
+     * Adds new gift certificate.
+     * <p>
+     * The certificate saved from <i>giftCertificateService</i>. Default tag "Main" is set for the certificate.
+     * If new tags are passed during creation certificate, they saved throw <i>tagService</i>Is returning
+     * <i>ResponseEntity</i> with <i>HttpStatus.CREATED</i>.
+     *
+     * @param giftCertificateDTO GiftCertificate instance.
+     * @param ucBuilder          UriComponentsBuilder instance.
+     * @return ResponseEntity.
+     */
+    @PostMapping(path = "/certificates", consumes = "application/json", produces = "application/json")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    public ResponseEntity<GiftCertificateDTO> addGiftCertificate(@RequestBody @Valid GiftCertificateDTO giftCertificateDTO,
+                                                                 UriComponentsBuilder ucBuilder) {
+        Long certificateId = giftCertificateService.saveCertificate(giftCertificateDTO);
+        HttpHeaders headers = new HttpHeaders();
+        headers.setLocation(ucBuilder.path("/certificates/{id}").buildAndExpand(certificateId).toUri());
+        return new ResponseEntity<>(headers, HttpStatus.CREATED);
```

Comment:
Return certificate object in response body. According to method signature result should be of ResponseEntity<GiftCertificateDTO> type